packages/@lwc/engine-dom/src/renderer.ts

```diff
@@ -143,211 +145,225 @@ if (isCustomElementRegistryAvailable()) {
-        return element.childNodes;
-    },
-
-    getFirstChild(element: Element): Node | null {
-        return element.firstChild;
-    },
-
-    getFirstElementChild(element: Element): Element | null {
-        return element.firstElementChild;
-    },
-
-    getLastChild(element: Element): Node | null {
-        return element.lastChild;
-    },
-
-    getLastElementChild(element: Element): Element | null {
-        return element.lastElementChild;
-    },
-
-    isConnected(node: Node): boolean {
-        return node.isConnected;
-    },
-
-    insertGlobalStylesheet(content: string): void {
-        if (!isUndefined(globalStylesheets[content])) {
-            return;
+export function attachShadow(element: Element, options: ShadowRootInit): ShadowRoot {
+    if (hydrating) {
+        return element.shadowRoot!;
+    }
+    return element.attachShadow(options);
+}
+
+export function setText(node: Node, content: string): void {
+    node.nodeValue = content;
+}
+
+export function getProperty(node: Node, key: string): any {
+    return (node as any)[key];
+}
+
+export function setProperty(node: Node, key: string, value: any): void {
+    if (process.env.NODE_ENV !== 'production') {
+        if (node instanceof Element && !(key in node)) {
+            // TODO [#1297]: Move this validation to the compiler
+            assert.fail(
+                `Unknown public property "${key}" of element <${
+                    node.tagName
+                }>. This is likely a typo on the corresponding attribute "${htmlPropertyToAttribute(
+                    key
+                )}".`
+            );
         }
+    }
 
-        globalStylesheets[content] = true;
+    (node as any)[key] = value;
+}
 
-        const elm = document.createElement('style');
-        elm.type = 'text/css';
-        elm.textContent = content;
+export function getAttribute(
+    element: Element,
+    name: string,
+    namespace?: string | null
+): string | null {
+    return isUndefined(namespace)
+        ? element.getAttribute(name)
+        : element.getAttributeNS(namespace, name);
+}
 
-        globalStylesheetsParentElement.appendChild(elm);
-    },
+export function setAttribute(
+    element: Element,
+    name: string,
+    value: string,
+    namespace?: string | null
+): void {
+    return isUndefined(namespace)
+        ? element.setAttribute(name, value)
+        : element.setAttributeNS(namespace, name, value);
+}
 
-    insertStylesheet(content: string, target: Node): void {
-        if (supportsConstructableStyleSheets) {
-            insertConstructableStyleSheet(content, target);
-        } else {
-            // Fall back to <style> element
-            insertStyleElement(content, target);
-        }
-    },
+export function removeAttribute(element: Element, name: string, namespace?: string | null): void {
+    if (isUndefined(namespace)) {
+        element.removeAttribute(name);
+    } else {
+        element.removeAttributeNS(namespace, name);
+    }
+}
```

Comment:
The `setAttribute`/`removeAttribute`/`getAttribute` APIs were inconsistent before:

https://github.com/salesforce/lwc/blob/43bad75ca088bee45509aacbe230a6ce9df7bcd8/packages/%40lwc/engine-core/src/framework/renderer.ts#L27

https://github.com/salesforce/lwc/blob/43bad75ca088bee45509aacbe230a6ce9df7bcd8/packages/%40lwc/engine-dom/src/renderer.ts#L225

https://github.com/salesforce/lwc/blob/43bad75ca088bee45509aacbe230a6ce9df7bcd8/packages/%40lwc/engine-server/src/renderer.ts#L230

I suppose technically we should be handling the `null` case for `namespace` in `@lwc/engine-dom`, but we weren't doing it before, so I kept it as-is. I made an effort not to actually change the implementation of the dom/server renderers anywhere; just the types.